lamagraph-core/src/Core/MemoryManager.hs

```diff
@@ -0,0 +1,383 @@
+              (actualAddr, actives, infoVec) = case sequenceA newLoadedNode of
+                Nothing -> (def, oldActives, accumulatedInfoMap)
+                Just localLoadedNode ->
+                  let addressIsActive = (newNodeIsActive . (^. containedNode) <$> localLoadedNode)
+                   in ( sequenceA $ Just $ (^. originalAddress) <$> localLoadedNode
+                      , mux addressIsActive ((replace . (^. originalAddress) <$> localLoadedNode) <*> addressIsActive <*> actives) actives
+                      , accumulatePortsChangesByLoadedNode accumulatedInfoMap localLoadedNode
+                      )
+              newLocalToActual = case sequenceA actualAddr of
+                Nothing -> localToActual
+                Just addr -> case sequenceA signalMaybeLocalNode of
+                  Nothing -> localToActual
+                  Just signalLocal -> updateLocalAddressToLoaded localToActual signalLocal addr
+           in (newBusyMap, newLocalToActual, actives, infoVec)
+      )
+      ( freedFromActivePair
+      , localActualMapDef
+      , pairsAfterDelete
+      , updatesInfo
+      )
+      localNodesSignal
+  -- accumulate all ports changes from edges. i.e. connect some external nodes with each other
+  infoAboutAllUpdates =
+    foldl
+      ( \infoVec signalMaybeEdge -> case sequenceA signalMaybeEdge of
+          Nothing -> infoVec
+          Just signalEdge -> updatePortsInfoByEdge <$> infoVec <*> signalEdge
+      )
+      infoAboutUpdatesByNodes
+      (unbundle $ (^. newEdges) <$> delta)
+  -- read all necessary external nodes from ram and update them with the local ones
+  nodesToWrite =
+    map
+      ( \signalMaybePair ->
+          case sequenceA signalMaybePair of
+            Nothing -> def -- same as @pure Nothing@
+            Just signalPair ->
+              let addr = fst <$> signalPair :: Signal dom AddressNumber
+                  signalMaybeInfo = snd <$> signalPair
+               in case sequenceA signalMaybeInfo of
+                    Nothing -> def
+                    Just signalInfo ->
+                      let localFlag = snd <$> signalInfo
+                          info = fst <$> signalInfo
+                          node = readByAddressAndFlag addr localFlag
+                       in mux
+                            (localFlag .==. pure External)
+                            (sequenceA $ Just $ LoadedNode <$> (updateNode <$> node <*> info) <*> addr)
+                            ( sequenceA $
+                                Just $
+                                  LoadedNode <$> (updateNode <$> node <*> info) <*> ((^. originalAddress) <$> localAddressToLoadedMap addr)
+                            )
+      )
+      (unbundle infoAboutAllUpdates)
+  -- write all changes into the ram
+  _ =
+    map
+      (maybe def writeByLoadedNode . sequenceA)
+      nodesToWrite
+  {-----------------
+  Helping functions
+  -----------------}
```

Comment:
If these functions do not use variables from the closure, maybe they can be moved to top-level and get correct Haddock documentation